torchtitan/experiments/kernels/moe/indices.py

```diff
@@ -203,6 +221,7 @@ def generate_permute_indices(
             tokens_per_expert_group,
             start_index_values,
             write_offsets,
+            total_tokens_per_expert,  # Pass to check for zero tokens
```

Comment:
We don't need to pass `total_tokens_per_expert` into the Triton kernel IMO.
The sole purpose of calculating `m_sizes` is just to get the `write_offsets`.
The kernel fills values by looking at `tokens_per_expert_group` (i.e. the drawing) and `start_index_values` (prefix sum of the former).

Reply:
We don't need to do this for correctness.  
This is a perf optimization to skip entire zero token experts and the associated memory loads from that.
Given that deepseek has 256 experts, we can assume a reasonable percentage of experts on every layer are not activated, thus I feel this check may have some performance gains.

Reply:
For training we still expect each expert to get activated

Reply:
If that were the case, users wouldn't have hit a hang here right :) 
( b/c there would not have been any zero token experts causing the hang, and thus need for padding in the first place.)
Anyway, since it seems both you and Ke are against this, I'll remove it b/c atm it's more important to just get this fix landed and move on to the training loop build out.